### PR TITLE
[V1] Initial support of multimodal models for V1 re-arch

### PR DESCRIPTION
**Pull request by ywang96**
This PR adds V1 support for selected image language models
- [x] Intern-VL
- [x] Molmo
- [x] Pixtral
- [x] ~Qwen2-VL (missing MRope implementation, will address in a separate PR)~ separated to #12128

### CODE REVIEW

File: vllm/model_executor/models/utils.py [hunk 1]
```diff
@@ -402,9 +402,17 @@ def merge_multimodal_embeddings(
     Note:
         This updates ``inputs_embeds`` in place.
     """
+    if isinstance(placeholder_token_id, int):
+        return _merge_multimodal_embeddings(
+            inputs_embeds,
+            (input_ids in placeholder_token_id),
```
Comment (DarkLight1337) [inline vllm/model_executor/models/utils.py:408]:
```suggestion
            (input_ids == placeholder_token_id),
```

I think you made a typo here?

File: vllm/model_executor/models/pixtral.py [hunk 2]
```diff
@@ -101,50 +103,80 @@ def input_mapper_for_pixtral(ctx: InputContext,
...
                                                     dtype=torch.float16)
         images.append(image)
+        image_tokens_list.append(encoding.tokens)
 
-    return MultiModalKwargs({"images": images})
+    image_tokens = torch.flatten(
+        torch.tensor([
+            token_id for image_tokens in image_tokens_list
+            for token_id in image_tokens
+        ]))
```
Comment (DarkLight1337) [inline vllm/model_executor/models/pixtral.py:133]: Is `flatten` necessary here? It looks like you're already doing that in the list comprehension.

File: vllm/model_executor/models/utils.py (hunk 1 above)
Comment (ywang96) [reply vllm/model_executor/models/utils.py:408]: Ah yes!

File: vllm/model_executor/models/pixtral.py (hunk 2 above)
Comment (ywang96) [reply vllm/model_executor/models/pixtral.py:133]: Good catch - initially I was debugging why it's returning 2D tensor and thought it's coming from here, then I realized it's from the `MultiModalKwargs.batch(mm_inputs)` call but forgot to update back. Let me remove it.

File: vllm/model_executor/models/molmo.py [hunk 3]
```diff
@@ -1210,7 +1223,16 @@ def get_input_embeddings(
...
         if multimodal_embeddings is not None:
-            inputs_embeds = inputs_embeds + multimodal_embeddings
+            assert isinstance(multimodal_embeddings, (list, tuple))
+            multimodal_embeddings = torch.cat(multimodal_embeddings, dim=0)
+            diff = inputs_embeds.shape[0] - multimodal_embeddings.shape[0]
+            if diff:
+                # We need to pad at the front of the multimodal embeddings to
+                # take input ids from other running requests into account.
+                inputs_embeds = inputs_embeds + F.pad(multimodal_embeddings,
+                                                      (0, 0, diff, 0))
```
Comment (ywang96) [inline vllm/model_executor/models/molmo.py:1233]: The padding logic here assumes that the decoding requests will be at the front of the batch. If this is not always the case please let me know! @WoosukKwon

**Review by DarkLight1337 — COMMENTED** [review]
Model changes look good to me overall. We can consider passing the placeholder indices to `get_input_embeddings` so that the relationship between `input_ids` and `multimodal_embeddings` becomes more explicit.

We should add tests to ensure that these models actually work in V1.

**Comment by ywang96** [comment]
> We can consider passing the placeholder indices to `get_input_embeddings` so that the relationship between `input_ids` and `multimodal_embeddings` becomes more explicit.

Hmm I don't think that alone will help actually because V1 assumes chunked prefill by default. Unless we're passing additional information such as computed tokens per request, the model runner doesn't know the original indices of `input_ids` in their corresponding requests.

File: vllm/v1/core/scheduler.py [hunk 4]
```diff
@@ -73,12 +73,12 @@ def __init__(
...
         # FIXME(woosuk): Below are placeholder values. We need to calculate the
         # actual values from the configurations.
-        self.max_num_encoder_input_tokens = 2048
+        self.max_num_encoder_input_tokens = 16384
         # NOTE(woosuk): For the models without encoder (e.g., text-only models),
         # the encoder cache will not be initialized and used, regardless of
         # the cache size. This is because the memory space for the encoder cache
         # is preallocated in the profiling run.
-        self.encoder_cache_manager = EncoderCacheManager(cache_size=2048)
+        self.encoder_cache_manager = EncoderCacheManager(cache_size=16384)
```
Comment (ywang96) [inline vllm/v1/core/scheduler.py:81]: These two numbers are raise to 16384 to serve all models included in this PR, otherwise a request can be stuck if its image has more tokens than 2048.

Eventually this number should be set based on model config.

File: vllm/model_executor/models/molmo.py (hunk 3 above)
Comment (WoosukKwon) [reply vllm/model_executor/models/molmo.py:1233]: As we discussed in the slack, this is not always the case 😢

Comment (ywang96) [reply vllm/model_executor/models/molmo.py:1233]: Hmm yea, then I wonder how to proceed from here - we probably need to take a look and see what information we can get from model runner to make this work.

Comment (ywang96) [reply vllm/model_executor/models/molmo.py:1233]: Actually - I will check their input processor code on HF to see if I can find anything useful

Comment (ywang96) [reply vllm/model_executor/models/molmo.py]: @WoosukKwon This has been resolved by [0176b7b](https://github.com/vllm-project/vllm/pull/10699/commits/0176b7b4618423cbe120a85db91ef4c9b4d28255)

File: vllm/v1/engine/llm_engine.py [hunk 5]
```diff
@@ -169,5 +174,18 @@ def start_profile(self):
     def stop_profile(self):
         self.engine_core.profile(False)
 
-    def get_tokenizer_group(self, group_type):
-        pass
+    def get_tokenizer_group(
```
Comment (WoosukKwon) [inline vllm/v1/engine/llm_engine.py:177]: Why is this needed for this PR?

File: vllm/model_executor/models/interfaces.py [hunk 6]
```diff
@@ -36,6 +36,11 @@ def get_multimodal_embeddings(self, **kwargs) -> Optional[T]:
         """
         Returns multimodal embeddings generated from multimodal kwargs 
         to be merged with text embeddings.
+
+        The output embeddings must be one of the following formats:
+        - A list or tuple of 2D tensors, where each tensor corresponds to 
+          each input image.
+        - A single 3D tensor, with the batch dimension grouping the 2D tensors.
```
Comment (WoosukKwon) [inline vllm/model_executor/models/interfaces.py:43]: Does this align with the type annotation?

File: vllm/multimodal/inputs.py [hunk 7]
```diff
@@ -96,7 +96,8 @@ class PlaceholderRange(TypedDict):
     """The length of the placeholder."""
 
 
-NestedTensors = Union[List["NestedTensors"], List[torch.Tensor], torch.Tensor]
+NestedTensors = Union[List["NestedTensors"], List[torch.Tensor], torch.Tensor,
+                      Tuple[torch.Tensor]]
```
Comment (WoosukKwon) [inline vllm/multimodal/inputs.py:100]: What does `Tuple[torch.Tensor]` mean here? Is it really a tuple with a single tensor?

File: vllm/model_executor/models/utils.py [hunk 8]
```diff
@@ -416,6 +416,15 @@ def merge_multimodal_embeddings(
     Note:
         This updates ``inputs_embeds`` in place.
     """
+    if isinstance(placeholder_token_id, list):
```
Comment (WoosukKwon) [inline vllm/model_executor/models/utils.py:419]: A dumb question: In which case could this be a list? Can you add comment somewhere?

File: vllm/v1/engine/llm_engine.py (hunk 5 above)
Comment (ywang96) [reply vllm/v1/engine/llm_engine.py:177]: This is needed for `llm.chat` interface as Pixtral referenced using this interface on their documentation.

File: vllm/model_executor/models/interfaces.py (hunk 6 above)
Comment (ywang96) [reply vllm/model_executor/models/interfaces.py:43]: See my other comment - will fix the Tuple typehinting issue.

File: vllm/model_executor/models/utils.py (hunk 8 above)
Comment (ywang96) [reply vllm/model_executor/models/utils.py:419]: Will do

File: vllm/model_executor/models/interfaces.py (hunk 6 above)
Comment (ywang96) [reply vllm/model_executor/models/interfaces.py:43]: Fixed

File: vllm/multimodal/inputs.py (hunk 7 above)
Comment (ywang96) [reply vllm/multimodal/inputs.py]: Fixed

File: vllm/model_executor/models/utils.py (hunk 8 above)
Comment (ywang96) [reply vllm/model_executor/models/utils.py:436]: I added a documentation in [610e662](https://github.com/vllm-project/vllm/pull/10699/commits/610e662d8a4d42827c5fdf71dbca2a6dd1287f30). Can you take a look and see if it makes sense?

File: vllm/model_executor/models/utils.py [hunk 9]
```diff
@@ -412,6 +412,23 @@ def merge_multimodal_embeddings(
     Merge ``multimodal_embeddings`` into ``inputs_embeds`` by overwriting the
     positions in ``inputs_embeds`` corresponding to placeholder tokens in
     ``input_ids``.
+    
+    ``placeholder_token_id`` can be a list of token ids (e.g, token ids
```
Comment (WoosukKwon) [inline vllm/model_executor/models/utils.py:416]: Oh does this mean that `S`, `B` and `E` (as well as `I`) come from the multi-modal embeddings rather than the vocab embeddings in the language model?

Comment (ywang96) [reply vllm/model_executor/models/utils.py:416]: `S`, `B` and `E`'s embeddings will come from vocab embeddings of the language model, the embeddings for `I` come from the visione encoder - we basically patch the multimodal embeddings with some text embeddings so that the entire thing can be slice-assigned into the text embeddings.

Comment (ywang96) [reply vllm/model_executor/models/utils.py:416]: Feel free to take a look at the changes in the pixtral.py (line 235 - 250) as an example

Comment (WoosukKwon) [reply vllm/model_executor/models/utils.py:416]: Got it. Thanks for the explanation.

File: vllm/v1/engine/llm_engine.py (hunk 5 above)
Comment (WoosukKwon) [reply vllm/v1/engine/llm_engine.py:177]: A dumb question: In which process does the tokenizer exist?

Comment (ywang96) [reply vllm/v1/engine/llm_engine.py:177]: Stealing the chart from https://github.com/vllm-project/vllm/pull/9826 for reference, the tokenizer lives in P0 since all inputs have been processed when sending to P1 so tokenizer shouldn't be needed there.

**Review by WoosukKwon — APPROVED** [review]
LGTM! It'd be nice if @DarkLight1337 can also take another look.

**Comment by ywang96** [comment]
It seems that we have test related to using image embeddings as input (which will be a breaking change in V1 since we cannot take batched image embedding directly as input), thus I'm separating Qwen2VL from this PR.

**Comment by DarkLight1337** [comment]
> LGTM! It'd be nice if @DarkLight1337 can also take another look.

Sorry for the late reply, it looks good to me.
